### DISCUSSION TITLE
Is it possible for Avalonia to train an AI to convert libraries such as av libglesv2, libHarfBuzzSharp, and libSkiaSharp into efficient pure C# implementations?

### DISCUSSION THREAD

**Discussion by sgf** [AvaloniaUI/Avalonia#15818] [Ideas]
Artificial intelligence and machine learning are so highly regarded these days. Isn’t it possible to do such a simple code translation?
Is there any chance for Avalonia to train an AI to convert C++ code to pure C#?
For example av_libglesv2,libHarfBuzzSharp,libSkiaSharp.

In this way, we can get rid of these dll dependencies, and we can use some AOT cropping functions to further reduce the application size.

Obviously, complex and repetitive tasks such as code conversion are best left to AI.

**Comment by wieslawsoltes** [comment]
@sgf  You need huge context window to make it work, and even then the code is probably not working with thousands of error. I think you have not used AI for any complex task. It fails miserably. AI is dumb.

**Reply by sgf** [reply]
I've tried, using ChatGPT to convert the code snippet.
Only need to post a piece of C/C++ code and tell him to convert it to C#.
It can be done. But as you said, its current capabilities are not enough to make a more perfect conversion.
I think this is due to lack of training.
In fact, it should be relatively easy to go from C/C++ to C#, but the other way around is more difficult. Because C/C++ may involve a large number of macros and conditional compilation.

However, I believe that as long as you continue to train and correct, it should be able to achieve a relatively perfect conversion.

**Reply by wieslawsoltes** [reply]
Code snippet is like comparing heating up water droplet to ocean, you might be surprised by the scale

**Reply by thevortexcloud** [reply]
Despite it's name, AI has no intelligence. It's a prediction engine. You input data, and it will output a prediction result based on the input. The GPT algorithm is not even that complicated by some standards, people have implemented it in only 500 or so lines of pure SQL.

https://explainextended.com/2023/12/31/happy-new-year-15/

In any case, I am not sure why you would even want to do this, much less use AI for it. AI is not the right tool for this. You would be better off using/writing a transpiler. EG Unity has one for converting MSIL (the code C# gets compiled into) into C++.

https://docs.unity3d.com/Manual/IL2CPP.html

AI can't and wont solve every problem, and in many cases it will create many new problems by people thinking it can solve every problem. 

You also can't remove every native dependency either, since at some point you will need to talk to a native library as the native word is written in C++/C/Rust/some other languages. EG GPU drivers, or a kernel API.  On Windows native libraries are generally shipped as DLL with the program. But I also don't see Windows using the Linux approach of having shared system libraries any time soon either (that has its own problems which the Windows world is not even close to being ready to solve). If you really want to get rid of them, I would suggest writing a different back end for Avalonia or using a different library/language entirely. You don't _have_ to use Skia with Avalonia either, it's just the best supported backend. There was someone awhile back that was even working on making Avalonia work in terminals using text output.

https://github.com/jinek/Consolonia/

Also even if you did manage to convert all the code, there is no guarantee it would result in any real advantages.

**Reply by sgf** [reply]
> Code snippet is like comparing heating up water droplet to ocean, you might be surprised by the scale

Only a few months have passed, and AI's token context processing capabilities have been greatly improved.
At present, some AI can handle up to 4 million tokens(every time). 
Conventional AI can also handle 1 million tokens. I believe this will happen soon, 
but AI’s ability to understand and translate code needs to be improved. Of course, 
there is also the ability to use programming tools. Currently, 
AI can only read and write code by force (in a sense, it is splicing code fragments that it has been trained on).
Of course, code bases like Skia are currently very large. Even 4 million is far from enough. 
Let us wait for the further development of AI. Maybe when there are 100 million tokens, it will not be a problem.

**Comment by kekekeks** [comment]
Converting С/C++ to C# is generally a bad idea performance-wise, especially on WASM.
In fact, we are considering to rewrite parts of Avalonia in native code.

**Reply by sgf** [reply]
In terms of writing regular code, the performance of C# is definitely not comparable to C++/rust. But if use a lot of Span<T>/pointers/structures, I believe its performance will not be worse than C++/rust. A A typical case is Microsoft's recently launched garnet (Redis alternative)

**Reply by robloo** [reply]
> In fact, we are considering to rewrite parts of Avalonia in native code.

Hm... in desktop scenarios (running on the CLR) C# is in the same order of magnitude of speed as C++. A lot of C# libraries are heavily optimized and are even faster than other language alternatives. If you drop into unsafe code you can do basically anything. So my first instinct is to say please don't switch languages for any core components of Avalonia. It is one of the most productive, easy-to-use languages and there are many developers that can contribute to a C# code base. Not true with C++ (WinUI3 is a worse-case example).

That said, I know interop ends up being the limiter in a lot of cases (Android). I also know that .NET browser WASM runtimes are still highly inefficient and aren't where they need to be. I do expect Microsoft to keep investing there though. Converting to C++ doesn't seem like it would help the WASM story either but I know next to nothing about web tech in this area.

**Reply by CastelloBrancoTecnologia** [reply]
> Converting С/C++ to C# is generally a bad idea performance-wise, especially on WASM. In fact, we are considering to rewrite parts of Avalonia in native code.

Bether than converting parte to c# is to implement an r3mobjects c# pipeline using island that produces native code comparable to c++ native code.

Avalonia in c# remobiects island has no interop to java on android and in win32 is already x64 native (or arm64) also in macos is equivalent as objective c naitive generated code and also in Linux is equivalent as native c generated code.

**Comment by danipen** [comment]
> In fact, we are considering to rewrite parts of Avalonia in native code.

Interesting... Which parts?

**Reply by kekekeks** [reply]
All of pixel manipulation, since C# code processing millions of pixels is **horrible** with wasm in debug mode (like 100x+ slower) and is moderately bad with conventional platforms (just 4-5x slower).

Also, C# interop story with certain C constructs like setjmp/longjmp is non-existant (for a good reason), so we can't directly utilize libs like libpng/libjpeg, which prevents us from writing image codecs that would handle extra metadata like EXIF,  DPI information, etc.

And maybe parts of the compositor around V13, because I'm kinda tired of C# not having proper mechanisms for automatic resource management.

**Comment by ALIENQuake** [comment]
@sgf For Avalonia Windows, you can use native aot static linking feature: https://github.com/AvaloniaUI/Avalonia/issues/9503

**Reply by sgf** [reply]
The size is too large. In China,some companies have trimed Skia on a large scale in order to use Skia, but unfortunately, they are not open source.
At Google, Flutter directly abandoned Skia and created Impeller.

**Reply by maxkatz6** [reply]
Skia abandonment is greatly exaggerated.
Google Chrome still uses Skia, and they rewrote it to use the new Skia Graphite API, equivalent if not better than Impeller.
At Avalonia we are considering using Skia Graphite, when it's stabilized. But it's not planned soon.

**Reply by maxkatz6** [reply]
And either way, both skia and impeller are native libraries that have to be shipped with .NET app. Or statically linked.
A fully managed GPU rendering solution is just not reasonable with existing options.
